### ISSUE TITLE
EntityType and ValueObjectType spawning/interring not working properly with Serializable interface (or other parent interfaces)

### ISSUE THREAD

**Issue by philippeboyd** [javers/javers#601]
So I've encountered a weird bug (or feature?) in Javers. The method `ReflectionUtil.calculateHierarchyDistance()` calculates the distance to be used in `TypeMapperState.findNearestAncestor()`. However, in the latter method there's this check `if(distances.get(0).isMax())` that will allow a null value to be returned by the method to be then used as an `Optional.ofNullable(prototype)` in the `TypeMapperState.infer` which will cause the `prototype` to NOT be present in the `TypeFactory.infer()` method which will then allow the javatype to be passed to inferFromAnnotations and eventually end up as an `EntityType` and not a `ValueObjectType`.

Here is my class hierarchy:

```
AbstractPermission -> Permission
AbstractPermission -> Role
```

Here are the logs :

**With Serializable interface on parent class**
```
o.j.core.metamodel.type.TypeFactory      : javersType of interface java.io.Serializable inferred as ValueObjectType

o.j.core.metamodel.type.TypeFactory      : javersType of class org.javers.mongo.javersmongoproblem.domain.Permission inferred as EntityType
o.j.core.metamodel.type.TypeFactory      : javersType of class java.lang.String inferred as ValueType, it's used as id-property type

o.j.core.metamodel.type.TypeFactory      : javersType of java.util.Set<java.lang.String> spawned as SetType from prototype SetType{baseType:interface java.util.Set}

o.j.core.metamodel.type.TypeFactory      : javersType of class org.javers.mongo.javersmongoproblem.domain.Role spawned as ValueObjectType from prototype ValueObjectType{baseType:interface java.io.Serializable}

o.j.core.metamodel.type.TypeFactory      : javersType of java.util.Set<org.javers.mongo.javersmongoproblem.domain.Permission> spawned as SetType from prototype SetType{baseType:java.util.Set<java.lang.String>}
```

**Without Serializable interface on parent class**
```
o.j.core.metamodel.type.TypeFactory      : javersType of interface java.io.Serializable inferred as ValueObjectType

o.j.core.metamodel.type.TypeFactory      : javersType of class org.javers.mongo.javersmongoproblem.domain.Permission inferred as EntityType
o.j.core.metamodel.type.TypeFactory      : javersType of class java.lang.String inferred as ValueType, it's used as id-property type

o.j.core.metamodel.type.TypeFactory      : javersType of java.util.Set<java.lang.String> spawned as SetType from prototype SetType{baseType:interface java.util.Set}

o.j.core.metamodel.type.TypeFactory      : javersType of class org.javers.mongo.javersmongoproblem.domain.Role inferred as EntityType
o.j.core.metamodel.type.TypeFactory      : javersType of class java.lang.String inferred as ValueType, it's used as id-property type

o.j.core.metamodel.type.TypeFactory      : javersType of java.util.Set<org.javers.mongo.javersmongoproblem.domain.Permission> spawned as SetType from prototype SetType{baseType:java.util.Set<java.lang.String>}
```

When AbstractPermission doesn't implement Serializable both object are known as EntityType during.

When AbstractPermission implements Serializable, Role object will get `spawned as ValueObjectType from prototype ValueObjectType{baseType:interface java.io.Serializable}` and will get the following error when getting changes

```java
QueryBuilder jqlQuery2 = QueryBuilder.byInstanceId(permissionReadUser.getId(), Permission.class);
List<Change> changes2 = javers.findChanges(jqlQuery2.build());
```

Will throw a : 
```
Caused by: org.javers.common.exception.JaversException: MANAGED_CLASS_MAPPING_ERROR: given javaClass 'class org.javers.mongo.javersmongoproblem.domain.Permission' is mapped to ValueObjectType, expected EntityType
```

It's a really complex problem and I'm pretty sure `ReflectionUtil.calculateHierarchyDistance()` isn't calculating something right...

You can test my case with my repository https://github.com/philippeboyd/javers-issue-managed-type

**Comment by bartoszwalacik** [comment]
Interesting case. Thanks for providing the runnable test. I will take a loot at this.

**Comment by philippeboyd** [comment]
I just found a workaround. Instead of letting Javers creating those Managed Type, I copied the Javers bean from the JaversMongoAutoConfiguration class and explicitly registered my Classes like so :

```java
@DependsOn("mongoTemplate")
@Bean
public Javers javers(MongoTemplate mongoTemplate, JaversProperties javersProperties) {
	return JaversBuilder.javers()
		.withListCompareAlgorithm(ListCompareAlgorithm.valueOf(javersProperties.getAlgorithm().toUpperCase()))
		.withMappingStyle(MappingStyle.valueOf(javersProperties.getMappingStyle().toUpperCase()))
		.withNewObjectsSnapshot(javersProperties.isNewObjectSnapshot())
		.withPrettyPrint(javersProperties.isPrettyPrint())
		.withTypeSafeValues(javersProperties.isTypeSafeValues())
		.registerJaversRepository(new MongoRepository(mongoTemplate.getDb()))
		.withPackagesToScan(javersProperties.getPackagesToScan())
		.registerEntities(Role.class, Permission.class) // <-- added line
		.build();
}
```

I'll leave this open for you @bartoszwalacik if you want to dig a little deeper.

**Comment by bartoszwalacik** [comment]
It's a good workaround, you can register your types explicitly but type inferring should also work. I will check why it doesn't.

**Comment by bartoszwalacik** [comment]
fix released in 3.7.0
